rfcs/45-stageless.md

```diff
@@ -0,0 +1,871 @@
+}
+```
+
+Labels are non-atomic by default, and atomic ordering constraints create coherent atomic groups.
+Isolated atomic groups are relatively niche, used only when you want to be certain that no information about internal progress can leak out.
+
+Calling `a.atomically_before(b)` will create a strict ordering constraint between `a` and `b`, and add them to the same atomic group (with a uniquely-generated system label).
+
+Of course, atomic groups (and thus atomic ordering constraints) should be used thoughtfully.
+As the strictest of the three types of system ordering dependency, they can easily result in unsatisfiable schedules if applied to large groups of systems at once.
+For example, consider the simple case, where we have an atomic ordering constraint from system A to system B.
+If for any reason commands must be flushed between systems A and system B, we cannot satisfy the schedule.
+The reason for this is quite simple: the lock on data from system A will not be released when it completes, but we cannot run exclusive systems unless all locks on the `World's` data are clear.
+As we add more of these atomic constraints (particularly if they share a single ancestor), the risk of these problems grows.
+In effect, atomically-connected systems look "almost like" a single system from the scheduler's perspective.
+
+On the other hand, atomic ordering constraints can be helpful when attempting to split large complex systems into multiple parts.
+By guaranteeing that the initial state of your complex system cannot be altered before the later parts of the system are complete,
+you can safely parallelize the rest of the work into separate systems, improving both performance and maintainability.
+
+### States
+
+**States** are one particularly common and powerful form of run criteria, allowing you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three sets of systems:
+
+1. **Update systems:** these systems run each schedule iteration if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.run_in_state(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(generate_map.run_on_enter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified state is exited.
+   1. `app.add_system(autosave.run_on_exit(GameState::Playing))`
+
+Update systems are by far the simplest: they're simply powered by run criteria.

```

Comment:
I'd replace "on update" with "on active" (there are a couple more mentions later in the doc).

Reply:
Used "while-active" instead; I think it's a bit more grammatical, and differentiates the behavior better.